Storage/persister/sherlock.h

```diff
@@ -71,44 +71,56 @@ class SherlockStreamPersisterImpl {
+        stream_(std::move(stream)),
+        publisher_used_(stream_->BecomeFollowingStream()) {
+    subscriber_instance_ = std::make_unique<SherlockSubscriber>(
+        [this](const transaction_t& transaction, std::chrono::microseconds timestamp) {
+          std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+          ApplyMutationsFromLockedSectionOrConstructor(transaction, timestamp);
+        });
+    std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+    SyncReplayStreamFromLockedSectionOrConstructor(0u);
   }
 
-  // TODO(dkorolev): `ScopeOwnedBySomeoneElse<>` ?
-  explicit SherlockStreamPersisterImpl(std::mutex& storage_mutex,
-                                       fields_update_function_t f,
-                                       sherlock_t& stream_owned_by_someone_else)
-      : storage_mutex_ref_(storage_mutex), fields_update_f_(f), stream_used_(stream_owned_by_someone_else) {
-    authority_ = (stream_used_.DataAuthority() == current::sherlock::StreamDataAuthority::Own)
-                     ? PersisterDataAuthority::Own
-                     : PersisterDataAuthority::External;
-    subscriber_ =
-        std::make_unique<SherlockSubscriber>([this](const transaction_t& transaction) { ApplyMutations(transaction); });
-    if (authority_ == PersisterDataAuthority::Own) {
-      // Do not use lock since we are in ctor.
-      SyncReplayStream<current::locks::MutexLockStatus::AlreadyLocked>();
-    } else {
-      SubscribeToStream();
-    }
+  SherlockStreamPersisterImpl(Following, fields_update_function_t f, Borrowed<stream_t> stream)
+      : fields_update_f_(f),
+        stream_publishing_mutex_ref_(stream->Impl()->publishing_mutex),
+        stream_(std::move(stream)) {
+    subscriber_instance_ = std::make_unique<SherlockSubscriber>(
+        [this](const transaction_t& transaction, std::chrono::microseconds timestamp) {
+          std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+          ApplyMutationsFromLockedSectionOrConstructor(transaction, timestamp);
+        });
+    std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+    SubscribeToStreamFromLockedSection();
   }
 
-  ~SherlockStreamPersisterImpl() { TerminateStreamSubscription(); }
+  ~SherlockStreamPersisterImpl() {
+    std::lock_guard<std::mutex> master_follower_change_lock(master_follower_change_mutex_);
+    TerminateStreamSubscriptionFromLockedSection();
+  }
 
-  PersisterDataAuthority DataAuthority() const {
-    std::lock_guard<std::mutex> lock(storage_mutex_ref_);
-    return authority_;
+  template <current::locks::MutexLockStatus MLS>
+  bool IsMasterStoragePersister() const {
+    locks::SmartMutexLockGuard<MLS> master_follower_change_lock(master_follower_change_mutex_);
+    return Exists(publisher_used_);
   }
 
-  void PersistJournal(MutationJournal& journal) {
+  template <current::locks::MutexLockStatus MLS>
+  std::chrono::microseconds UpToDateUntilPersister() const {
```

Comment:
Maybe just `LastAppliedTimestamp`? Why introduce different names for a single concept?